sdk-lib/src/main/java/cash/z/ecc/android/sdk/SdkSynchronizer.kt

```diff
@@ -194,15 +202,49 @@ class SdkSynchronizer private constructor(
         }
     }
 
-    private val _status = MutableStateFlow<Synchronizer.Status>(DISCONNECTED)
-    private val _exchangeRateUsd = MutableStateFlow<Pair<BigDecimal, Instant>?>(null)
+    private val _status = MutableStateFlow(DISCONNECTED)
 
-    var coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
+    val coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
 
     override val orchardBalances = processor.orchardBalances.asStateFlow()
     override val saplingBalances = processor.saplingBalances.asStateFlow()
     override val transparentBalance = processor.transparentBalance.asStateFlow()
-    override val exchangeRateUsd = _exchangeRateUsd.asStateFlow()
+
+    private val refreshExchangeRateUsd = MutableSharedFlow<Unit>(replay = 1).apply { tryEmit(Unit) }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override val exchangeRateUsd =
+        channelFlow {
+            var lastValue = ObserveFiatCurrencyResult()
+            refreshExchangeRateUsd
+                .flatMapLatest {
+                    flow {
+                        emit(lastValue.copy(isLoading = true))
+                        lastValue =
+                            when (val result = fetchExchangeChangeUsd()) {
+                                is FetchFiatCurrencyResult.Error -> lastValue.copy(isLoading = false)
```

Comment:
Is this implementing a partial copy? i.e. if we have older valid data, and then a refresh ends in an error, we preserve that older data (so queries to the `StateFlow` still return an exchange rate)?

Reply:
exactly